### PR TITLE
[extractor/DouyuTV]: Fix extractor and use new api

### PR DESCRIPTION
**Pull request by hatienl0i261299**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

ADD DESCRIPTION HERE

Fixes #

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

I got some issue with old API like this:
image

And i updated to new hash and API.

### CODE REVIEW

**Comment by pukkandan** [comment]
Why should I spend time reviewing this when you are closing multiple PRs without addressing them?

File: yt_dlp/extractor/douyutv.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 
 class DouyuTVIE(InfoExtractor):
     IE_DESC = '斗鱼'
-    _VALID_URL = r'https?://(?:www\.)?douyu(?:tv)?\.com/(?:[^/]+/)*(?P<id>[A-Za-z0-9]+)'
+    _VALID_URL = r'https?://(?:www\.)?douyu(?:tv)?\.com/(topic/.*rid=|(?:[^/]+/))*(?P<id>[A-Za-z0-9]+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/douyutv.py:16]: Don't use `.*`. The regex should not match any of these
```
https://www.douyu.com/topic/rid=6560603
https://www.douyu.com/topic/ydxc#rid=6560603
https://www.douyu.com/topic/ydxc?drid=6560603
```

File: yt_dlp/extractor/douyutv.py [hunk 2]
```diff
@@ -53,13 +53,29 @@ class DouyuTVIE(InfoExtractor):
...
         },
         'params': {
             'skip_download': True,
         },
+    }, {
+        'url': 'https://www.douyu.com/topic/ydxc?rid=6560603',
+        'info_dict': {
+            'id': '6560603',
+            'display_id': '6560603',
+            'ext': 'm3u8',
```
Comment (pukkandan) [inline yt_dlp/extractor/douyutv.py:68]: Wrong

File: yt_dlp/extractor/douyutv.py [hunk 3]
```diff
@@ -53,13 +53,29 @@ class DouyuTVIE(InfoExtractor):
...
+            'title': str,
+            'description': str,
```
Comment (pukkandan) [inline yt_dlp/extractor/douyutv.py:70]: See other tests

File: yt_dlp/extractor/douyutv.py [hunk 4]
```diff
@@ -53,13 +53,29 @@ class DouyuTVIE(InfoExtractor):
...
+            'thumbnail': r're:^https?://.*\.png',
+            'uploader': '阿涛皎月Carry',
+            'live_status': 'is_live',
+            'is_live': True,
```
Comment (pukkandan) [inline yt_dlp/extractor/douyutv.py:74]:
```suggestion
```

File: yt_dlp/extractor/douyutv.py [hunk 5]
```diff
@@ -79,28 +95,25 @@ def _real_extract(self, url):
             room_id = self._html_search_regex(
                 r'"room_id\\?"\s*:\s*(\d+),', page, 'room id')
 
-        # Grab metadata from mobile API
+        # Grab metadata from API
+        tt = int(time.time())
+        auth = md5(f'room/{video_id}?aid=wp&client_sys=wp&time={tt}zNzMV1y4EMxOHS6I5WKm'.encode('utf-8')).hexdigest()
```
Comment (pukkandan) [inline yt_dlp/extractor/douyutv.py:100]:
```suggestion
        auth = md5(f'room/{video_id}?aid=wp&client_sys=wp&time={tt}zNzMV1y4EMxOHS6I5WKm'.encode()).hexdigest()
```

File: yt_dlp/extractor/douyutv.py [hunk 6]
```diff
@@ -79,28 +95,25 @@ def _real_extract(self, url):
...
-        video_url = self._download_json(
-            'http://coapi.douyucdn.cn/lapi/live/thirdPart/getPlay/' + room_id,
-            video_id, note='Downloading video URL info',
-            query={'rate': 0}, headers={
-                'auth': sign,
-                'time': str(tt),
-                'aid': 'pcclient'
-            })['data']['live_url']
+        has_url = room.get('hls_url')
+        video_url = 'https://hls3-akm.douyucdn.cn/' + self._search_regex(r'(live/.*)', has_url, 'Extract has_url')
```
Comment (pukkandan) [inline yt_dlp/extractor/douyutv.py:115]:
```suggestion
        video_url = url_join('https://hls3-akm.douyucdn.cn/',
                             self._search_regex(r'(live/.*)', room['hls_url'], 'URL')
```

File: yt_dlp/extractor/douyutv.py [hunk 7]
```diff
@@ -79,28 +95,25 @@ def _real_extract(self, url):
...
+        tt = int(time.time())
+        auth = md5(f'room/{video_id}?aid=wp&client_sys=wp&time={tt}zNzMV1y4EMxOHS6I5WKm'.encode('utf-8')).hexdigest()
         room = self._download_json(
-            'http://m.douyu.com/html5/live?roomId=%s' % room_id, video_id,
-            note='Downloading room info')['data']
+            f'http://www.douyutv.com/api/v1/room/{room_id}', video_id,
+            note='Downloading room info', query={
+                'aid': 'wp',
+                'client_sys': 'wp',
+                'time': tt,
```
Comment (pukkandan) [inline yt_dlp/extractor/douyutv.py:106]: duplicate code. Why not construct the query object first and use it to generate auth. Like this:
https://github.com/yt-dlp/yt-dlp/blob/074b2fae9076221faaa8697381428131ad968dc9/yt_dlp/extractor/kankanews.py#L29-L38

File: yt_dlp/extractor/douyutv.py [hunk 8]
```diff
@@ -1,5 +1,5 @@
 import time
-import hashlib
+from hashlib import md5
```
Comment (pukkandan) [inline yt_dlp/extractor/douyutv.py:2]: revert

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/douyutv.py [hunk 9]
```diff
@@ -13,7 +14,7 @@
 
 class DouyuTVIE(InfoExtractor):
     IE_DESC = '斗鱼'
-    _VALID_URL = r'https?://(?:www\.)?douyu(?:tv)?\.com/(?:[^/]+/)*(?P<id>[A-Za-z0-9]+)'
+    _VALID_URL = r'https?://(?:www\.)?douyu(?:tv)?\.com/(topic/\w+\?rid=|(?:[^/]+/))*(?P<id>[A-Za-z0-9]+)/?$'
```
Comment (pukkandan) [inline yt_dlp/extractor/douyutv.py:17]: Why the `$`?

Comment (hatienl0i261299) [reply yt_dlp/extractor/douyutv.py]: Its not necessary, i removed it.

**Comment by suger520HT** [comment]
How to modify douyutv.py to make "https://hls3-akm.douyucdn.cn/live/592541r8sShiTTrP/playlist.m3u8?wsSecret=76803a83bfa4d7281e33955d2dd201bb&wsTime=1683694929, 137 "Transformation" https://openhls-tct .douyucdn2.cn/ Live broadcast/16101rumZjzVOZu1.m3u8?wsSecret=5ead1ab74a99de9f865906e94c50e14c&wsTime=1683787603", hla3a-akm cannot be used in Mainland China

**Comment by bashonly** [comment]
@suger520HT
```diff
diff --git a/yt_dlp/extractor/douyutv.py b/yt_dlp/extractor/douyutv.py
index fa40844df..c138d6fb0 100644
--- a/yt_dlp/extractor/douyutv.py
+++ b/yt_dlp/extractor/douyutv.py
@@ -111,7 +111,7 @@ def _real_extract(self, url):
         if room.get('show_status') == '2':
             raise ExtractorError('Live stream is offline', expected=True)
 
-        video_url = urljoin('https://hls3-akm.douyucdn.cn/', self._search_regex(r'(live/.*)', room['hls_url'], 'URL'))
+        video_url = urljoin('https://openhls-tct.douyucdn2.cn/', self._search_regex(r'(live/.*)', room['hls_url'], 'URL'))
         formats, subs = self._extract_m3u8_formats_and_subtitles(video_url, room_id)
 
         title = unescapeHTML(room['room_name'])
```

**Comment by suger520HT** [comment]
Can you remove /playlist in "https://openhls-tct/ .douyucdn2.cn/ Live broadcast/16101rumZjzVOZu1/playlist.m3u8?wsSecret=5ead1ab74a99de9f865906e94c50e14c&wsTime=1683787603"

**Comment by suger520HT** [comment]
> @suger520HT
> 
> ```diff
> diff --git a/yt_dlp/extractor/douyutv.py b/yt_dlp/extractor/douyutv.py
> index fa40844df..c138d6fb0 100644
> --- a/yt_dlp/extractor/douyutv.py
> +++ b/yt_dlp/extractor/douyutv.py
> @@ -111,7 +111,7 @@ def _real_extract(self, url):
>          if room.get('show_status') == '2':
>              raise ExtractorError('Live stream is offline', expected=True)
>  
> -        video_url = urljoin('https://hls3-akm.douyucdn.cn/', self._search_regex(r'(live/.*)', room['hls_url'], 'URL'))
> +        video_url = urljoin('https://openhls-tct.douyucdn2.cn/', self._search_regex(r'(live/.*)', room['hls_url'], 'URL'))
>          formats, subs = self._extract_m3u8_formats_and_subtitles(video_url, room_id)
>  
>          title = unescapeHTML(room['room_name'])
> ```

Can you remove /playlist in "https://openhls-tct/ .douyucdn2.cn/ Live broadcast/16101rumZjzVOZu1/playlist.m3u8?wsSecret=5ead1ab74a99de9f865906e94c50e14c&wsTime=1683787603"

**Comment by suger520HT** [comment]
> @suger520HT
> 
> ```diff
> diff --git a/yt_dlp/extractor/douyutv.py b/yt_dlp/extractor/douyutv.py
> index fa40844df..c138d6fb0 100644
> --- a/yt_dlp/extractor/douyutv.py
> +++ b/yt_dlp/extractor/douyutv.py
> @@ -111,7 +111,7 @@ def _real_extract(self, url):
>          if room.get('show_status') == '2':
>              raise ExtractorError('Live stream is offline', expected=True)
>  
> -        video_url = urljoin('https://hls3-akm.douyucdn.cn/', self._search_regex(r'(live/.*)', room['hls_url'], 'URL'))
> +        video_url = urljoin('https://openhls-tct.douyucdn2.cn/', self._search_regex(r'(live/.*)', room['hls_url'], 'URL'))
>          formats, subs = self._extract_m3u8_formats_and_subtitles(video_url, room_id)
>  
>          title = unescapeHTML(room['room_name'])
> ```

Can you remove the /playlist in "https://openhls-tct/ .douyucdn2.cn/ Live /16101rumZjzVOZu1/playlist.m3u8?wsSecret=5ead1ab74a99de9f865906e94c50e14c&wsTime=1683787603", only Can you remove /playlist in "https:// openhls- tct/ .douyucdn2.cn/ Live /16101rumZjzVOZu1.m3u8?wsSecret=5ead1ab74a99de9f865906e94c50e14c&wsTime=1683787603" to play with /playlist will not play

**Comment by suger520HT** [comment]
image
You need to delete the /playlist, and then replace hls3a-akm.douyucdn.cn with openhls-tct.douyucdn.cn to play it, otherwise the parsed link is /playlist.m3u8....it will not play

**Comment by bashonly** [comment]
@suger520HT see edit

**Comment by suger520HT** [comment]
thank you

**Comment by suger520HT** [comment]
image
Problem solved, thank you very much sir

**Comment by gamer191** [comment]
@bashonly do you think that patch should be merged into master?

**Comment by bashonly** [comment]
I don't know enough about the site or the accessibility of these HLS URLs to make that call

**Comment by suger520HT** [comment]
I don't know, I only know that hls3a-akm cannot use this domain name in mainland China

> 我对网站或这些HLS URL的可访问性不够了解，无法进行该调用

**Comment by suger520HT** [comment]
> I don't know enough about the site or the accessibility of these HLS URLs to make that call

I don't know,
hls3a-akm breaks after a few seconds of playback in mainland China

**Comment by suger520HT** [comment]
> I don't know enough about the site or the accessibility of these HLS URLs to make that call

Using a proxy and then using hls3a-akm will not interrupt the playback after a few seconds

**Comment by suger520HT** [comment]
Can you set conditions for hls3a-akm? If it can’t play, switch to openhls-tct.douyucdn2.cn

**Comment by suger520HT** [comment]
> @suger520HT
> 
> ```diff
> diff --git a/yt_dlp/extractor/douyutv.py b/yt_dlp/extractor/douyutv.py
> index fa40844df..79621a2ac 100644
> --- a/yt_dlp/extractor/douyutv.py
> +++ b/yt_dlp/extractor/douyutv.py
> @@ -111,8 +111,8 @@ def _real_extract(self, url):
>          if room.get('show_status') == '2':
>              raise ExtractorError('Live stream is offline', expected=True)
>  
> -        video_url = urljoin('https://hls3-akm.douyucdn.cn/', self._search_regex(r'(live/.*)', room['hls_url'], 'URL'))
> -        formats, subs = self._extract_m3u8_formats_and_subtitles(video_url, room_id)
> +        video_url = urljoin('https://openhls-tct.douyucdn2.cn/', self._search_regex(r'(live/.*)', room['hls_url'], 'URL'))
> +        formats, subs = self._extract_m3u8_formats_and_subtitles(video_url.replace('/playlist.m3u8', '.m3u8'), room_id)
>  
>          title = unescapeHTML(room['room_name'])
>          description = room.get('show_details')
> ```

https:// openhls-tct/ .douyucdn2.cn/live/16101rumZjzVOZu1.flv?wsSecret=5ead1ab74a99de9f865906e94c50e14c&wsTime=1683787603 Change to output.flv? The effect is better m3u8 will eat and upload flv but not

**Comment by suger520HT** [comment]
> @suger520HT
> 
> ```diff
> diff --git a/yt_dlp/extractor/douyutv.py b/yt_dlp/extractor/douyutv.py
> index fa40844df..79621a2ac 100644
> --- a/yt_dlp/extractor/douyutv.py
> +++ b/yt_dlp/extractor/douyutv.py
> @@ -111,8 +111,8 @@ def _real_extract(self, url):
>          if room.get('show_status') == '2':
>              raise ExtractorError('Live stream is offline', expected=True)
>  
> -        video_url = urljoin('https://hls3-akm.douyucdn.cn/', self._search_regex(r'(live/.*)', room['hls_url'], 'URL'))
> -        formats, subs = self._extract_m3u8_formats_and_subtitles(video_url, room_id)
> +        video_url = urljoin('https://openhls-tct.douyucdn2.cn/', self._search_regex(r'(live/.*)', room['hls_url'], 'URL'))
> +        formats, subs = self._extract_m3u8_formats_and_subtitles(video_url.replace('/playlist.m3u8', '.m3u8'), room_id)
>  
>          title = unescapeHTML(room['room_name'])
>          description = room.get('show_details')
> ```
Ask if you can generate a streamlined link http://openhls-tct.douyucdn2.cn/live/252140rz7aIq14cx.flv?uuid=

**Comment by suger520HT** [comment]
> @suger520HT
> 
> ```diff
> diff --git a/yt_dlp/extractor/douyutv.py b/yt_dlp/extractor/douyutv.py
> index fa40844df..79621a2ac 100644
> --- a/yt_dlp/extractor/douyutv.py
> +++ b/yt_dlp/extractor/douyutv.py
> @@ -111,8 +111,8 @@ def _real_extract(self, url):
>          if room.get('show_status') == '2':
>              raise ExtractorError('Live stream is offline', expected=True)
>  
> -        video_url = urljoin('https://hls3-akm.douyucdn.cn/', self._search_regex(r'(live/.*)', room['hls_url'], 'URL'))
> -        formats, subs = self._extract_m3u8_formats_and_subtitles(video_url, room_id)
> +        video_url = urljoin('https://openhls-tct.douyucdn2.cn/', self._search_regex(r'(live/.*)', room['hls_url'], 'URL'))
> +        formats, subs = self._extract_m3u8_formats_and_subtitles(video_url.replace('/playlist.m3u8', '.m3u8'), room_id)
>  
>          title = unescapeHTML(room['room_name'])
>          description = room.get('show_details')
> ```

Ask if you can generate a streamlined link http://openhls-tct.douyucdn2.cn/live/252140rz7aIq14cx.flv?uuid=

**Comment by suger520HT** [comment]
douyu的那个失效了 能修复一下吗

**Comment by suger520HT** [comment]
http://openhls-tct.douyucdn2.cn/dyliveflv1a/5168755rYvNEHQvI.m3u8?txSecret=ed532144e99db67a9cb97d81f56d01e2&txTime=648725f3
现在的有效格式

**Pull request merged**